server/selection_test.go

```diff
@@ -0,0 +1,250 @@
+	if r.err != nil {
+		return nil, r.err
+	}
+	return r.stakes, nil
+}
+
+func (r *stubStakeReader) SetStakes(stakes map[ethcommon.Address]*big.Int) {
+	r.stakes = stakes
+}
+
+func TestSessHeap(t *testing.T) {
+	assert := assert.New(t)
+
+	h := &sessHeap{}
+	heap.Init(h)
+	assert.Zero(h.Len())
+	// Return nil for empty heap
+	assert.Nil(h.Peek())
+
+	sess1 := &BroadcastSession{LatencyScore: 1.0}
+	heap.Push(h, sess1)
+	assert.Equal(h.Len(), 1)
+	assert.Equal(h.Peek().(*BroadcastSession), sess1)
+
+	sess2 := &BroadcastSession{LatencyScore: 1.1}
+	heap.Push(h, sess2)
+	assert.Equal(h.Len(), 2)
+	assert.Equal(h.Peek().(*BroadcastSession), sess1)
+
+	sess3 := &BroadcastSession{LatencyScore: .9}
+	heap.Push(h, sess3)
+	assert.Equal(h.Len(), 3)
+	assert.Equal(h.Peek().(*BroadcastSession), sess3)
+
+	assert.Equal(heap.Pop(h).(*BroadcastSession), sess3)
+	assert.Equal(heap.Pop(h).(*BroadcastSession), sess1)
+	assert.Equal(heap.Pop(h).(*BroadcastSession), sess2)
+	assert.Zero(h.Len())
+}
+
+func TestMinLSSelector(t *testing.T) {
+	assert := assert.New(t)
+
+	sel := NewMinLSSelector(nil, 1.0)
+	assert.Zero(sel.Size())
+
+	sessions := []*BroadcastSession{
+		&BroadcastSession{},
+		&BroadcastSession{},
+		&BroadcastSession{},
+	}
+
+	sel.Add(sessions)
+	assert.Equal(sel.Size(), 3)
+	for _, sess := range sessions {
+		assert.Contains(sel.unknownSessList, sess)
+	}
+
+	// Select from unknownSessList
+	sess1 := sel.Select()
+	assert.Equal(sel.Size(), 2)
+	assert.Equal(len(sel.unknownSessList), 2)
+
+	// Set sess1.LatencyScore to not be good enough
+	sess1.LatencyScore = 1.1
+	sel.Complete(sess1)
+	assert.Equal(sel.Size(), 3)
+	assert.Equal(len(sel.unknownSessList), 2)
+	assert.Equal(sel.knownSessList.Len(), 1)
+
+	// Select from unknownSessList
+	sess2 := sel.Select()
+	assert.Equal(sel.Size(), 2)
+	assert.Equal(len(sel.unknownSessList), 1)
+	assert.Equal(sel.knownSessList.Len(), 1)
+
+	// Set sess2.LatencyScore to be good enough
+	sess2.LatencyScore = .9
+	sel.Complete(sess2)
+	assert.Equal(sel.Size(), 3)
+	assert.Equal(len(sel.unknownSessList), 1)
+	assert.Equal(sel.knownSessList.Len(), 2)
+
+	// Select from knownSessList
+	knownSess := sel.Select()
+	assert.Equal(sel.Size(), 2)
+	assert.Equal(len(sel.unknownSessList), 1)
+	assert.Equal(sel.knownSessList.Len(), 1)
+	assert.Equal(knownSess, sess2)
+
+	sel.Clear()
+	assert.Zero(sel.Size())
+	assert.Nil(sel.unknownSessList)
+	assert.Zero(sel.knownSessList.Len())
+	assert.Nil(sel.stakeRdr)
+}
+
+func TestMinLSSelector_UnknownSessListSelect_Errors(t *testing.T) {
+	assert := assert.New(t)
+
+	stakeRdr := newStubStakeReader()
+	sel := NewMinLSSelector(stakeRdr, 1.0)
+
+	sel.Add(
+		[]*BroadcastSession{
+			&BroadcastSession{
+				OrchestratorInfo: &net.OrchestratorInfo{
+					TicketParams: &net.TicketParams{Recipient: []byte("foo")},
+				},
+			},
+		},
+	)
+
+	// Test error when reading stake
+	stakeRdr.err = errors.New("Stakes error")
+	assert.Nil(sel.unknownSessListSelect())
```

Comment:
Can we add a check that an error was logged using `glog.Stats.Error.Lines()`?

Reply:
Fixed in 703441e